### ISSUE TITLE
Discussion: show Git-ignored files by default?

### ISSUE THREAD

**Issue by sharkdp** [sharkdp/fd#612]
Since `fd` was first published, the feature to hide Git-ignored files by default has always been controversial. It's the number one pitfall for new users, as witnessed by the numerous issues that have been opened over time (even though this is the first point in the [Troubleshooting](https://github.com/sharkdp/fd#troubleshooting) section). Even experienced users will likely run into this from time to time.

We have had past discussions about this (see #179), but I'm not so sure anymore if this default is the best possible option for the "average user".

I thought it might make sense to discuss this again and see what others think. Whatever we choose as the default, it will always be easy for users to select a different default via an alias.

**Pro** "*ignore `.gitignore` entries by default*" (current default):
- Most searches are faster if we take `.gitignore` files into account. `.gitignore`d directories tend to contain huge amounts of automatically generated build artifacts or downloaded dependency files. Pruning these directories from the search tree typically results in a faster search overall. There are counterexamples to this where the parsing of long `.gitignore` files takes longer than actually traversing these directories.
- Most of the time, `.gitignore`d results are not "interesting" to the user (however, see counterpart below).
- When running `fd` without any arguments, I typically don't want to see `.gitignore`d files.

**Cons**:
- It can be very confusing to (new) users. If 10% of users go so far as to create a ticket on GitHub to ask about their problem, there must be hundreds of users that ran into this problem at some point.
- Even if you know about the default, it can be annoying to repeat the search because you forgot to add `-I` or `-u`. There are a lot of valid use cases where users are - in fact - interested in results from ignored directories or files. Personally, I would estimate that I use `-uu` or `-HI` in roughly 20% of my searches, which is quite high.

**Comment by kpym** [comment]
I want to add that the nature of the files in `.gitignore` depends a lot on the nature of our projects. In my case for example most of the time the ignored files are files with sensitive information (not crap) that I want to be able to search with `fd`.

But I understand that for other people often the files in `.gitignore` have to be ignored by `fd` as well.

For this reason the desired default behavior is not the same for everyone.

In my opinion, the default behavior should be "search all", because it's easier to figure out why there are too many results than it is to figure out why there are missing results. 

But such a change in behavior will not be backward compatible, which is never good. To overcome this, people must be allowed to easily return to the old default behavior. Hence the need to be able to configure the default behavior of `fd` (#362).

**Comment by sharkdp** [comment]
> because it's easier to figure out why there are too many results than it is to figure out why there are missing results.

I think this is an excellent point.

> But such a change in behavior will not be backward compatible, which is never good.

Agreed. This is probably the main reason why I never changed the behavior. Still, if it should turn out that 90% of users would like a different default, I'm more than willing to make a breaking change.



> To overcome this, people must be allowed to easily return to the old default behavior. Hence the need to be able to configure the default behavior of `fd` (#362).

Okay, I have reopened the ticket once again. Let's discuss this aspect in #362. There are other ways to configure the default behavior as well (aliases, wrappers, environment variables).

**Comment by tmccombs** [comment]
Personally, I think the pros outweigh the cons. But setting up an alias is easy, so I wouldn't be too upset if this changed.

**Comment by Weker01** [comment]
I am a new user and think this is a cool feature but I also think it should be the default as it is not obvious from any short description and is not intuitive. 

"Powerusers" could easily setup an alias as was commented earlier.

**Comment by jonathan-s** [comment]
I would add my vote to search all files except hidden files by default.

**Comment by et342** [comment]
Hello.

I'd like to make a point that `fd` is a general-purpose file-searching utility that is not git specific, so having it to take into account `.gitignore` files, laying around in the filesystem, does not feel right.
In fact, I've stumbled upon this issue the very first time I tried `fd`: I've tried to find something, starting from a non-git directory in subdirectories which happened to be git repos and found nothing, although I knew it was there. After that, the very next thing I did is patched `fd` locally, so it wouldn't read `.gitignore` files by default.

**Comment by kpym** [comment]
## The dilemma

I think that for this issue and for #362 the question is:

**Should `fd` behaves as "general" or "git-style" tool ?**

What I mean by this is summarized in the following table

|               | general tool  |           git style tool            |
|:--------------|:-------------:|:-----------------------------------:|
| examples      | `find`,`grep` |             `git`,`rg`              |
| ignore files  |      no       |                 yes                 |
| configuration |  flags only   | flags, files, environment variables |

Actually `fd` is in between the two worlds. And respecting the ignore files without the configuration feature is bad, IMO. So `fd` should choose the red pill or the blue pill ;)

## My proposal

Or may be we don't have to choose and we can have both.

I think that :

- `fd` should be a **general tool** (by default)
- there should be a flag `--git` that makes it act in a "git style"
- `fdg` should be another compiled version from the same code but with different default behavior, that is equivalent to `fd --git` and heaving `--no-git` flag to switch it to "general tool".

My arguments for creating the additional `fdg` are the following:

- it should be easy to make the automation tool to build two versions, one with `--no-git` option as default, the other with `--git` as default.
- the alias solution discussed already in #362 is very "shell specific" :
  - there is no general alias that can work in all shells (`powershell` and `cmd` included)
  - every time we download `fd` we should setup the aliases, so the advantage of "no install, just download and use it" is not valid any more
  - aliases are not working when you wan to make a general script (let's say in python) using `fd`
- this solve the backward compatibility with ignore files as any user can decide simply to switch to `fdg` or even to rename `fdg` fo `fd` and continue using it as before (respect the ignore files by default)
- this allows to solve not only this issue but also # and all other issues that ask to change the default behavior of `fd` (this would be done using the configuration file in `fdg`)
- having both versions will make the supporters on both sides ("pro-general" and the "pro-git") happy

**Comment by tmccombs** [comment]
I like that idea, with one caveat. Instead of having a separately compiled version of fd, fdg should just be a symlink to fd, and fd check the name that it is called with, and if it is "fd" use the general behavior, and if it is "fdg" use the "git" behavior. Or alternatively distribute OS-specific wrapper scripts for fdg (for example that does something like `exec fd --ignore`, or whatever the windows equivalent is).

**Comment by kpym** [comment]
@tmccombs  The problem with all workarounds : aliases, scripts, symlinks, ... are the same : 
- they are all system/shell specific
- in general they are ok in the linux world but not in windows
- you should configure all this before to start using the tool
- they can generate additional problems that will create more issues here

Did you have arguments against heaving `fdg` in addition ? Once the CI/CD tool configured to compile both tools (for all systems) this do not create more work for the maintainers.

The only drawback I see is that having two versions can confuse the novice user which one to take. But this can be easily solved by promoting `fd` and talking about `fdg` in a section at the end of the README.

**Comment by tmccombs** [comment]
Sorry, I am suggesting that the symlink or shell would be distributed with fd. Having separate, but nearly identical binaries means the package is larger (longer to download and more space on disk). fd is small enough it's probably not that big of a deal for most people, but it still feels wrong to me. 

> Once the CI/CD tool configured to compile both tools (for all systems) this do not create more work for the maintainers.

It means builds take longer, which affects anyone who builds it from source.

**Comment by kpym** [comment]
> Sorry, I am suggesting that the symlink or shell would be distributed with fd. 

This means that all scripts should be tested and maintened and they will raise more issues here.
I have very bad experience with the provided scripts for windows for example with `conda` (python package manager).
Providing additional scripts is the end of "single executable tool".

> Having separate, but nearly identical binaries means the package is larger (longer to download and more space on disk). fd is small enough it's probably not that big of a deal for most people, but it still feels wrong to me.

I don,t know what you call "the package". The builds are larger, but the sources are not. And this argument is valid for all os specific builds. If we want smallers builds it is easy : provide only one (or zero) builds and tell the peopel to build their own, but this is not very user firnedly.

> > Once the CI/CD tool configured to compile both tools (for all systems) this do not create more work for the maintainers.
> 
> It means builds take longer, which affects anyone who builds it from source.

No. You build only the version that you need. This is already the case for all os specific builds.
Only the CI/CD builds all the versions.

And if we have to choose only one type of build (general vs git style) which one this should be ?
Giving a good solution only for one category of users and telling to the other category "get by with scripts" is not very user friendly, IMO.

**Comment by tmccombs** [comment]
> This means that all scripts should be tested and maintened and they will raise more issues here.
The script I'm proposing is extremely simple, it would just call fd with an option and the arguments passed to the script. If we had a separate executable, that would have to be tested and maintained too :).

> I don,t know what you call "the package".

I mean the zip, tarball, or deb you download from the [releases page](https://github.com/sharkdp/fd/releases), or a package you install with a package manager like apt, dnf, pacman, homebrew, winget (maybe?), etc. 

> Providing additional scripts is the end of "single executable tool".

From your next paragraph, I think what you mean by this is that you need to install more than just a single executable. But afaik, this is not a design goal of fd. And fd is not currently distributed as just an executable. all packages include command completions scripts, and on linux they also include man pages. 

> The builds are larger, but the sources are not. And this argument is valid for all os specific builds. If we want smallers builds it is easy : provide only one (or zero) builds and tell the peopel to build their own, but this is not very user firnedly.

So you're saying we would have twice as many packages on the release page and in package managers/app stores? I really don't like that idea. I think it would make it even more confusing for users to know which package to install/download, and it roughly doubles the amount of work for packagers to maintain the package for package managers like apt, dnf, pacman, etc. 

> No. You build only the version that you need.

You're assuming no one would want both. This also means complicates building from source, since the user now needs to be able to configure which version they want to build somehow (the OS they want is implicit in the OS they are running the build from, unless they are cross-building).

> And if we have to choose only one type of build (general vs git style) which one this should be ?
Giving a good solution only for one category of users and telling to the other category "get by with scripts" is not very user friendly, IMO.

That is not what I'm suggesting.

If using a script really is such a problem for windows, why not the symlink/command name approach?

fd would have something like:

```rust
let cmdName = std::env::args().next().unwrap();
if cmdName.ends_with("fdg") {
  // use git-style default
} else {
  // use "general" default
}
```

Then on unix-like systems, the package would just have a symlink from fdg to fd (or vice versa). If windows doesn't have an equivalent it could just have a copy of fd called fdg, or have a seperate package if that makes more sense, and/or have instructions to rename the file in the readme depending on what functionality you want.

**Comment by ftpd** [comment]
We already have ~/.fdignore. Maybe this file could somehow 'include' gitignore (via something like _@~/.gitignore_ or some other character/directive)? With this approach showing git-ignored files could be enabled via default, also allowing user to add his git-ignored entries that are already in _~/.gitignore_ (or _./.git/ignore_ when inside repository) in an easy way?

Personally, I wasn't even aware that git-ignored files are omitted: https://imgur.com/a/UlLD8ED
For now my _.fdignore_ contains mostly 100% of _.gitignore_ + other patterns. It would be great to be able to 'include' the file as a whole, not to copy it's content manually.

**Comment by tmccombs** [comment]
> Maybe this file could somehow 'include' gitignore (via something like @~/.gitignore or some other character/directive)

respecting .gitignore is more than just respecting a global ~/.gitignore file though. It's also uses .gitignore files in ancestor directories, and descendent directories (scoped to those directories).   You _could_ have some syntax in .fdignore to enable `--ignore-vcs`, but that isn't really an include, so much as a flag.  And we would still probably want a way to disable that with `--no-ignore-vcs` on the command line to overrule that, but keep the other ignores from .fdignore, which seems kind of weird to me.

**Comment by joshuarli** [comment]
FWIW, ripgrep (which I imagine a lot of fd users also use) also respects gitignore by default.

**Comment by et342** [comment]
Thought dump:

`git` is primarily concerned about the contents of files, their state, but not their presence.
This means that `.gitignore` files are also about the state of the files, but not their presence.

`ripgrep`, just like `git`, also primarily concerned about the contents of files,
and this shared concern makes its choice of consideration of `.gitignore` files understandable,
although it could also be opt-in.

`fd`, on the other hand, is not concerned about the state of the files, but is concerned about their presence,
what differs from concerns of `git` and `ripgrep`, what makes its consideration of `.gitignore` files slightly less fitting.

**Comment by joshuarli** [comment]
> `fd`, on the other hand, is not concerned about the state of the files, but is concerned about their presence

That's a good way of looking at it, my stance is pretty neutral now. As a long-time user I'd have no problem making an inverse alias to `fdh: aliased to fd --no-ignore-vcs -H -E '.git/'`, but I still feel there's value in performant (I assume most fd users have git repos in their filesystems) defaults. Plus, it's already established and would take some effort to flip.

**Comment by elihunter173** [comment]
Just to throw my opinion into the ring. I'm in favor of changing the default.

When I use fd with no options/arguments other than a pattern, 99% of the time I'm just using it to quickly narrow down the list of files I need to look at to find what I want. In that case I'm okay if I get some things that I don't care about in the search, but I'm much more annoyed if I don't find something that's actually there because I forgot to specify that I wanted to search .gitignored files as well. @sharkdp said that he adds `-H` or `-u` around 20% of the time, meaning those flags mattered 20% of the time. But I'm willing to bet that if those flags were enabled by default then they would have to be disabled much less than 20% of the time.

Also, from a scripting/reducing noise perspective, normally when when I'm doing something more precise than just quickly narrowing down a list of files, I'm more willing to add flags and check the documentation in order to narrow down the search results to be only what I care about.

And concerning adding an `fdg` binary (or symlink), I don't see how that's better than just adding an optional flag. It feels like it would complicate CI and packaging a lot for something that essentially just flips a flag on by default.

**Comment by matu3ba** [comment]
I have an additive suggestion, which could leverage or make the suggestion obsolete: Add an according description to [tldr](https://github.com/tldr-pages/tldr).
`rg`/`ripgrep` has a description `rg -uu pattern`, which is the second result and thus searchable in 1s.

20% typing the thing would then overall still mean less time.
Bonus is that `-uu` could be established as *use hidden github stuff* or "do more work".

**Comment by kvelicka** [comment]
Some thoughts on a couple of points brought up in this thread, though nothing terribly new.

1. Conflating _git_ ignore with general ignore

I'm also in favour of changing the default, because IMO paths being present `.gitignore` mean literally what it says on the tin - "not interesting _for the purposes of version control_" and I wish that tools like `fd` as well as `ripgrep` did not overload this definition to mean roughly "not interesting to search/scan in by default". I, like many others here, have had false negatives due to this. It's (subjectively!) a bit sad that in order to be confident in a negative search result one has to either provide extra flags, or rerun the search using a "legacy" tool.

2. Special treatment of `.gitignore` with regards to other similar files

`fd`, or to be more precise the `ignore` crate that it uses, appears to only support `git`'s ignore files, which means `fd`'s behaviour for say Mercurial users will be different. Firefox, arguably the poster child for Rust, uses Mercurial for example.


3. Prior art re: aliases/separate binaries

This applies to `grep`, and is likely Linux (or perhaps even Debian) specific:

```
root@9fb4e89aea1b:/# man grep | head -4
GREP(1)                                                                User Commands                                                               GREP(1)

NAME
       grep, egrep, fgrep, rgrep - print lines that match patterns
```

So, using aliases for commonly-used flags is at the very least nothing new.

**Comment by matu3ba** [comment]
> Some thoughts on a couple of points brought up in this thread, though nothing terribly new.
> 
>     1. Conflating _git_ ignore with general ignore
> 
> 
> I'm also in favour of changing the default, because IMO paths being present `.gitignore` mean literally what it says on the tin - "not interesting _for the purposes of version control_" and I wish that tools like `fd` as well as `ripgrep` did not overload this definition to mean roughly "not interesting to search/scan in by default". I, like many others here, have had false negatives due to this. It's (subjectively!) a bit sad that in order to be confident in a negative search result one has to either provide extra flags, or rerun the search using a "legacy" tool.
> 
>     1. Special treatment of `.gitignore` with regards to other similar files
> 
> 

Did you ever run `grep` on repos with huge binary files (>5 GB) or big amount of files ignored by `.gitignore` ?
Especially binary data (without newline) use linear time and that is why `ripgrep` has another default than `grep`.
For usage for `fd` of many, many files inside `.gitignore` ie compiling Linux Kernel the same argument can be made.

> `fd`, or to be more precise the `ignore` crate that it uses, appears to only support `git`'s ignore files, which means `fd`'s behaviour for say Mercurial users will be different. Firefox, arguably the poster child for Rust, uses Mercurial for example.
> 

Argument of authority is no technical argument on usage. And you cant make everyone happy for using the tool. Here a short catalogue for decision making:

0. **Usage consistency** What type of consistency to other tools (arguments + effects) should be used? (for me that is Rust and ripgrep, if possible)?
1. **Usage purpose** Ignoring build files has the purpose of supporting dev environments, where you frequently want to search for (relative) filepaths in a complex tree. (effect of clear speed win as less paths and files need to be traversed)
2. **Oriented user base** If the author chooses to support such thing, when should a version control tool be supported? (market share, user base?) 
3. **Clarification** How should it be documented? (manual, cheat sheet, [tealdeer](https://github.com/dbrgn/tealdeer),[tldr](https://github.com/tldr-pages/tldr) )
4. **Technical feasability** How many build files can be "hosted" (as result of codegen) on Mercurial to justify ignoring them?
5. **Usage feasability** How many build files are "hosted" (as result of codegen) on Mercurial to justify ignoring them?

>     1. Prior art re: aliases/separate binaries
> 
> 
> This applies to `grep`, and is likely Linux (or perhaps even Debian) specific:
> 
> ```
> root@9fb4e89aea1b:/# man grep | head -4
> GREP(1)                                                                User Commands                                                               GREP(1)
> 
> NAME
>        grep, egrep, fgrep, rgrep - print lines that match patterns
> ```
> 
> So, using aliases for commonly-used flags is at the very least nothing new.

How should this be maintained and name-clashes prevented ? `cfdisk`, `df`, `efi`,`rfkill` are already used. Do you have specific names in mind?

**Comment by tmccombs** [comment]
> This applies to `grep`, and is likely Linux (or perhaps even Debian) specific:
> 
> ```
> root@9fb4e89aea1b:/# man grep | head -4
> GREP(1)                                                                User Commands                                                               GREP(1)
> 
> NAME
>        grep, egrep, fgrep, rgrep - print lines that match patterns
> ```
> 
> So, using aliases for commonly-used flags is at the very least nothing new.

On Ubuntu at least, egrep, fgrep and rgrep are simply shell scripts that run grep with the given options. I would be ok with that, or with a symlink approach. But I'd rather not have seperate but nearly identical compiled binaries.

**Comment by kvelicka** [comment]
(repo maintainer - this reply certainly cuts close to being off-topic, please feel free to remove it)

> Did you ever run `grep` on repos with huge binary files (>5 GB) or big amount of files ignored by `.gitignore` ?
> Especially binary data (without newline) use linear time and that is why `ripgrep` has another default than `grep`.
> For usage for `fd` of many, many files inside `.gitignore` ie compiling Linux Kernel the same argument can be made.

* grep ignores binaries by default as well
* most repos don't contain 5GB binaries, nor nearly as many build artifacts as the LInux kernel

I'd like to refer back to the original issue, which talks about picking the best default for the "average user", not about removing support for ignoring based on `.gitignore` altogether. Overall, the ability to piggyback on `.gitignore` is tremendously helpful. However, I have been bitten by it myself and have seen others in the same situation, hence my _personal_ stance of "don't look at `.gitignore` by default". It's an anecdotal account and you can absolutely find more valid critiques against it, but I don't think that's gonna advance the overall discussion much more.

> Argument of authority is no technical argument on usage. And you cant make everyone happy for using the tool. Here a short catalogue for decision making:

My argument (2) is definitely a weak one, and you're right that it's impossible to make everyone happy (e.g. look at us having this very discussion!). I mentioned it because `fd` (`ripgrep` is arguably more guilty, but that's offtopic) talks about version control ignore files in a general sense, but in fact supports only `git`. From `fd --help`: 

* "<...>that would otherwise be ignored by '.*ignore' files"
* the very name of `--no-ignore-vcs` flag

Again, this is only a quibble.

> How should this be maintained and name-clashes prevented ? `cfdisk`, `df`, `efi`,`rfkill` are already used. Do you have specific names in mind?

I don't have any specific suggestions here. My aim is only to highlight that using ~~aliases~~ shell scripts to invoke certain flags is something that's already ships with popular Linux distros.

**Comment by tmccombs** [comment]
> fd (ripgrep is arguably more guilty, but that's offtopic) talks about version control ignore files in a general sense, but in fact supports only git

I believe this is for forwards compatibility. So if at some future point the ignore crate adds support for additional VCS systems (such as mercurial), then the `--no-ignore-vcs` flag will just work, without having to add a new `--no-ignore-hg` flag and similar for each VCS system that is added.

**Comment by matu3ba** [comment]
>     * grep ignores binaries by default as well
> 
>     * most repos don't contain 5GB binaries, nor nearly as many build artifacts as the LInux kernel
> 

Not, when you did not write a binary-data header or alike.

> My argument (2) is definitely a weak one, and you're right that it's impossible to make everyone happy (e.g. look at us having this very discussion!). I mentioned it because `fd` (`ripgrep` is arguably more guilty, but that's offtopic) talks about version control ignore files in a general sense, but in fact supports only `git`. From `fd --help`:
> 
>     * "<...>that would otherwise be ignored by '.*ignore' files"
> 
>     * the very name of `--no-ignore-vcs` flag
> 
> 
> Again, this is only a quibble.

True.
@sharkdp 
Why is checking`.*ignore*` for files/folders to ignore not possible? The filepath needs to be checked against the `.gitignore` anyway or is there a limit to 1 or 2?

> I don't have any specific suggestions here. My aim is only to highlight that using ~aliases~ shell scripts to invoke certain flags is something that's already ships with popular Linux distros.

I dont like them at all and prefer shell aliases.

**Comment by tobiww** [comment]
If you do make this change, _please_ consider using separate flags for .gitignore, .fdignore, etc. I have run into valid use cases for (observe .fdignore, ignore .gitignore) and visa-versa.

Examples: 
`-I/--ignore -- Ignore file patterns in .gitignore and .fdignore`
`-Ig/--ignore-gitignore -- Ignore file patterns in .gitignore`
`-If/--ignore-fdignore -- Ignore file patterns in .fdignore`
`-N/--no-ignore -- do Not ignore file patterns in .gitignore and .fdignore`
`-Ng/--no-ignore-gitignore -- do Not ignore file patterns in .gitignore`
`-Nf/--no-ignore-fdignore -- do Not ignore file patterns in .fdignore`

Unfortunately these all use double-negatives, and there is a potential confusion about the double-meaning of "ignore .gitignore" (ignoring the .gitignore file and ignoring the files within it have opposite meaning). Other terms that may be less confusing:
- E/N - --Exclude-from  (exclude files listed in .{}ignore), do Not exclude
- U/N - Use, do Not use

There is precedent for fine-grained ignore params in ripgrep: 
(`--no-ignore-dot, --no-ignore-global, --no-ignore-vcs`, etc.)

[ If the above comment about supporting non-git repositories is implemented,  then `Ig` might become `Iv` (for vcs) ]

**Comment by tmccombs** [comment]
I agree with having more granular control over ignore files. However, I'd rather use flags consistent with what ripgrep uses, both because of consistentency and because I think they are a little less confusing.

**Comment by fregante** [comment]
I think the point of having a tool like this is that **it's opinionated.** If I have to start adding flags to reach the default behavior/length of `find`, I might as well use `find`

Like `rg` and the rest of the modern tools, what makes them great is their defaults. If the only advantage is a very minor speed bump, people would just use the preinstalled tools they already know.

**Comment by christianbundy** [comment]
Wow, thanks for pinning this issue. I manage my dotfiles by creating a git repo in `~` and then adding `*` to `~/.gitignore`. I was _so_ confused at why I wasn't getting any results in my home directory.

I think we should follow the convention from `grep` and `git grep`:

- `fd` does not exclude files from `.gitignore`
- `git fd` excludes files from `.gitignore`

**Comment by Weker01** [comment]
> Wow, thanks for pinning this issue. I manage my dotfiles by creating a git repo in `~` and then adding `*` to `~/.gitignore`. I was _so_ confused at why I wasn't getting any results in my home directory.
> 
> I think we should follow the convention from `grep` and `git grep`:
> 
>     * `fd` does not exclude files from `.gitignore`
> 
>     * `git fd` excludes files from `.gitignore`
> 
> 
> Workarounds: I need to make sure that any time `fd` is invoked (by any script!) it's called with `--no-ignore-vcs`. I'd love to be able to set this in a configuration file or environment variable.

This is not a convention, git grep is a subcommand of git while grep is a completely different programm.

**Comment by christianbundy** [comment]
> This is not a convention.

This *is* a convention -- the `git-grep` binary is provided by the Git project, but from the perspective of a user it's clear that `git-grep` takes into account `.gitignore` whereas `grep` does not. I'm suggesting that Git-specific usage should be provided under `git-fd` rather than `fd`. In summary:

- `foo` -- should aim to be as general-purpose as practical
- `git-foo` -- should aim to be as Git-specific as practical

>  `git grep` is a subcommand of git while `grep` is a completely different programm.

Correct, I don't think anyone is confused about that.

**Comment by tmccombs** [comment]
>     fd does not exclude files from .gitignore
>     git fd excludes files from .gitignore

I have a few issues with this:
1. `git grep` works very differently than `fd`. While `git grep` runs `grep` on each file that is checked in (basically equivalent to `git ls-files | xargs grep`), fd does file traversal itself, in parallel, which s where you get a lot of the speed. 
2. git subcommands are generally implied to be operating on a single git repository. But I sometimes want to use `fd` (and `rg`) in a directory that contains multiple git repos and still respect .gitignore files.   
3. It means adding a new executable that needs to be installed, and two levels of indirection when calling it.

In short, I don't think it is any better of an option than making `--no-ignore-vcs` the dfault rather than `--ignore-vcs`.

**Comment by aral** [comment]
In case it helps, I thought `fd` was broken while searching for something I knew was in my `node_modules` directory due to this.

**Comment by sergeevabc** [comment]
Windows 7 user here. I have a dedicated folder with CLI tools added to PATH variable. For example, there is `ripgrep` with `.ripgreprc` next to it, which contains settings I need with every launch like `--smart-case` or coloring preferences. I like the portability of this approach instead of polluting %UserProfile%, registry or creating more env variables. So it would be nice to have similar configuration here. I would use it to make ` -H` permanent, because I always forget to add it (`find` shows all).

**Comment by fats** [comment]
Just adding my experience here that this caught me off guard multiple times. Most users install fd as a replacement for find, so it can be surprising when it doesn't show certain files by default.

**Comment by sharkdp** [comment]
Adding this to the "fd 9" milestone because I would like to settle this discussion and introduce the (possible) breaking change in that version (see #613).

**Comment by tavianator** [comment]
My suggestion would be to not change the default, so we don't break anyone's workflow.  Instead, how about something like this:

```shell
$ fd -e o
(9 ignored files skipped, 2 hidden files skipped; see fd -h for details)
```

With #595 implemented, users could make `fd` an alias to `fd --no-hidden --ignore` to keep the current behaviour and suppress the warning, or to `fd -HI` to show everything by default.

I'd be okay with always printing that warning, even if there are matches.  But especially if there are no matches it might be handy.

**Comment by tmccombs** [comment]
If directories are skipped would it count those a 1?

**Comment by tavianator** [comment]
It would yeah.  Maybe the exact number isn't important though, how about "some ignored files and hidden files were skipped"?

**Comment by pemistahl** [comment]
I'm in favor of @tavianator's suggestion. For me, the current default makes perfect sense because I'm generally not interested in searching ignored files. The other mentioned proposals (two separate binaries, configuration files, a lot more flags) are, in my humble opinion, the opposite of _simple_ and _user-friendly_ ways to deal with this "problem". To me, it feels like breaking a butterfly on a wheel. Just let the user know if and how many ignored or hidden files have been skipped. Informative, straightforward, easy to implement.

**Comment by jonathan-s** [comment]
> Just let the user know if and how many ignored or hidden files have been skipped. 

I would also be in favour of this. Also inform the user how they can make sure _not_ to skip these files.

**Comment by sharkdp** [comment]
> easy to implement

Well :smile:. I'm not so sure about this. We currently use the `ignore` crate for parallel directory traversal + gitignore handling. I don't think it (currently) has a mode where it gives us all files, but marks that ones that would be ignored (or similar). But I might be wrong.

The bigger problem I see is with performance. If we want to show this warning, it would mean that we will always have to parse gitignore files, even in the presence of `-I`/`--no-ignore`. This can potentially result in significant performance regressions for (no-ignore) searches.

**Comment by tavianator** [comment]
> Well smile. I'm not so sure about this. We currently use the `ignore` crate for parallel directory traversal + gitignore handling. I don't think it (currently) has a mode where it gives us all files, but marks that ones that would be ignored (or similar). But I might be wrong.

Yeah, might require a patch to `ignore`.  We don't need to know what paths they were, or even how many, just whether it ignored anything.

> If we want to show this warning, it would mean that we will always have to parse gitignore files, even in the presence of `-I`/`--no-ignore`.

I don't think we need to show the warning when `-I` is passed, at least about ignored files.  We could still warn about hidden files unless `-H` is passed.

**Comment by sharkdp** [comment]
> I don't think we need to show the warning when `-I` is passed, at least about ignored files.

Oh - of course! In this case that's a route that we should explore :+1:

**Comment by tmccombs** [comment]
I like that idea as well.  Although for scripting you probably don't want that in the output.  But I think writing it to stderr would probably be ok, and stderr can be redirected to /dev/null if desired.

**Comment by mg979** [comment]
My 2 cents: if `fd` is advertized as an alternative to `find`, it should behave as `find` as closely as possible. I was initially confused as well.

99% percent of the times I want to ignore git files, but still I would prefer if `fd` searched everything by default. It's `fd` not `fdg` after all. The lack of command line options should default to the most general and inclusive behaviour, imho. It's how most command line tools behave I think.

For example I'd prefer something like:
```sh
fd .           # don't ignore anything, as find
fd . -i=fig    # respect .fdignore, .ignore and .gitignore
fd . -I        # respect all ignore files
```

**Comment by j-james** [comment]
Although I'm in favor of `fd` searching gitignored files by default, `fd` already makes a fairly radical break from `find` by switching the order of parameters.

**Comment by tavianator** [comment]
Another thing to add to [my suggestion](https://github.com/sharkdp/fd/issues/612#issuecomment-896049390), having just debugged #876: perhaps if errors occur during the search, that should be noted as well.  Something like

```
$ fd certbot.pem /etc
(17 errors occurred, results may be incomplete; pass --show-errors for details)
```

**Comment by jchook** [comment]
If this default changes, I would humbly request that CLI flags allow overrides to previous CLI flags.

For example...

```sh
# Will ignore 
fd --no-ignore --ignore

# Will not ignore
fd --ignore --no-ignore
```

This way folks can easily choose their default via an interactive shell alias, but still have the option e.g.

```sh
alias fd="fd --ignore"
```

**Comment by tavianator** [comment]
@jchook This should be done already by https://github.com/sharkdp/fd/pull/822

**Comment by andreavaccari** [comment]
I also like @tavianator proposal, possibly with the following caveats:

- `fd` should print a warning only if outputting to `stdout` in an interactive shell
- The warning should not include counts for performance reasons but simply state:

```
(Rerun with `-u` to include ignored files, `-uu` to include all files)
```

**Comment by dylan-tock** [comment]
> In case it helps, I thought `fd` was broken while searching for something I knew was in my `node_modules` directory due to this.

Ditto for me looking for an installed gem in `$HOME/.rbenv`

**Comment by anacrolix** [comment]
I believe the default should be changed. Similarly showing errors should be the default. Normally those things are suppressed when you have special needs or expectations with a search. When you run `fd` in frustration to work out where things are, you don't want things hidden from you.

**Comment by stevenwalton** [comment]
I want to add that I think this is a bad default and there are bad premises, probably based on the type of developer ***you*** are, not ***developers*** in general. 

> Most of the time, .gitignored results are not "interesting" to the user 

Most of the time files in `.gitignore` ***are*** the most interesting ones and exactly what I want to be searching for. If you work with compiled languages, systems with logs, or tasks that produce binaries then you are highly interested in these. You want to search your logs (and `fd` with `fzf` makes this great). For tasks with binaries you often want to confirm their creation or that they were accurately cleared. I want to search logs and have better tools than `grep` and be able to find these across swaths of directories (work HPC or ML and you'll be doing this a lot). But neither of these would be things you would push upstream and are going to be in your `gitignore`. 

The question of "interesting" depends on what kind of developer you are and the ***default*** should be the case with the ***least*** assumptions made. 

> It's the number one pitfall for new users, as witnessed by the numerous issues that have been opened over time **(even though this is the first point in the [Troubleshooting](https://github.com/sharkdp/fd#troubleshooting) section**). ***Even experienced users will likely run into this from time to time.***

This assumes that every user will read the documentation. I understand the frustration, I really do. But people are not going to hear about `fd` from the github but their friends, blogs, reddit, whatever. They'll do `apt get install fd-find` and are told this is a fancy alternative to `find`.  The unfortunate truth is that you could plaster this warning in big bold flashing letters at the top of the github page and you'd still have frequent user confusion. You can only directly control so much and you can't force users to read documentation [just as you can't expect users to read a pull/push sign](https://99percentinvisible.org/article/norman-doors-dont-know-whether-push-pull-blame-design/).

> A simple, fast and user-friendly alternative to 'find' 

The way it is advertised gives a reasonable user interpretation of expecting behavior to be like `find`. I understand that you've even updated the feature list to note this ignoring of files. BUT I think no matter how clear you make this, it is generally going to be unseen. The flag is no doubt useful just as I am a frequent user of the exclude glob patterns. 

An additional and critical point I want to make is that we end up getting different behavior based on where we issue our find command due to changing `gitignore`s. If I run this in the project directory I will get different results than if I run it in a different project directory that is a clone of all the non-git files. This is highly unexpected behavior and will rightfully make a user feel schizophrenic. 

Overall the question comes down to which is the worst case:
1. A user can't find a file they expected to because it is captured by `.gitignore`
2. A user gets additional output

I think 1 is a obviously a ***significantly*** worse situation. There's an adage in experimental sciences that applies: it is better to record too much information than too little. This is because too little has (generally) far higher costs than not enough (which requires doing everything over because you didn't know everything you'd need beforehand).

I write this message not to hate on this project but for the exact opposite reason. I absolutely love it, but I also want to see it become the best tool it can be.

**Comment by matu3ba** [comment]
> A user gets additional output

I do consider finding additional potential unwanted things differently from ripgrep a bigger problem than some missing ones. As example, if you are deleting files or you are probing untrusted files and do not want divergence of tooling for unwanted surprises.

Personally, I do use fd intentionally, because it is able to use expressed intent by git repo authors, for example to manage my dotfiles and it feels kinda weird not to do.
Personally I'd prefer one tool combining ripgrep and fd to make things simple and not learning yet another arbitrary default, but I do recognize that this is unfeasible to maintain.

**Comment by tmccombs** [comment]
> If you work with compiled languages, systems with logs, or tasks that produce binaries then you are highly interested in these.

I do work with compiled languages, systems with logs, and tasks that produce binaries. And when I use fd, I usually don't want to include those things. I want to be able to search for a source file, without showing files that are generated from it. Indeed, one of the biggest attractions of fd when I first started using it was that it respected gitignore by default. Although I will admit it isn't quite as important to me as it is for ripgrep. 

Unfortunately, it is hard to get accurate data on how many users would prefer one default over the other. We will get a bias in responses to this issue from people who don't like the current default. But I strongly suspect that if we changed it we would get a lot of complaints from people who want it changed back.

For interactive use, I'm starting to think having some way to configure defaults would be worthwhile. But I worry about the impact that would have on startup time, especially if we parse a config file. And for scripting, having the behavior depend on a users preferences is... not great. For the latter, perhaps we could mitigate with a flag that puts fd in "scripting mode" where we ignore whatever mechanism we use for configuration.

**Comment by stevenwalton** [comment]
`rg` is not part of the standard command set and isn't really relevant to this conversation. 

My argument isn't ***I*** (Steven) have this particular use case vs ***you*** (tmccombs or matu3ba) have a particular use case. I just gave those as an example.

My argument is "which default yields the lowest entropy"

The reasoning to my argument is "follow same set of defaults as the standard system." 

Personally, I'm just throwing `alias fd=fd --no-ignore` into my rc and calling it a day. From a design perspective, I strongly believe more confusion is created by a default that excludes files that one would produce via standard commands such `find`, `ls`, `grep`, `locate`, and so on. We're talking about default options. If you're using `rg` I assume you can be like me and throw `alias fd=fd --git` into your rc and call it a day too. The question is not "what do ***you*** find useful" but "what behavior is most expected from a new user". Let's just make sure we get the framing right and let's also not forget that `alias` exists. I mean we all have dotfiles, right? 

The frequent issues are explicit evidence that such default behavior does create huge surprisal to users. So is the fact that it's in the first line of the documentation and in the feature list. When it doesn't, those users probably read the documentation closely. If a user reads the documentation closely they can easily throw in an alias into their rc file (because that's what those files are for, personalization) and go on about their day and the github issues will go away. We can even think about this from another perspective if the several I have given aren't enough. Which would be a larger breaking change: if the default is to ignore the `*foo` glob and you remove that default filter or the default is to have no filter and you introduce a `*foo` glob. Obviously the latter results in a higher surprisal to the user. 

The arguments for the default filter are arguments of personal use case, which is why I said the desired behavior depends on what type of developer you are. Providing customization options are fantastic and I'm super happy `fd` has these. That still doesn't change the issue that the current default creates higher surprisal. If you want to convince the `--no-ignore` crowd that we're wrong you have to convince us that this default creates less surprisal.

You're not going to go over to `exa` or `lsd` and find tons of issues "command outputs files that pattern match gitignore, this is unexpected behavior." ***It would be silly to think so and that's why it feels weird to even be having this discussion.*** I am surprised that you are surprised that people are surprised that the "better find" tool filters out files that aren't hidden.

**Comment by kpym** [comment]
IMO, all of this discussion about what are the best defaults points to the following conclusions:
- **There are no perfect defaults**.
- There should be a way for everyone to set their own defaults, independent of the way they call `fd` (aliases are not, they are shell dependent), so having a config file is IMO mandatory to solve this issue (see #362).
- To make the `fd` runs reproducible, a `--ignore-config` flag should be added if the config file is introduced (see #362).

My personal opinion on **what are the "best" defaults should be discussed from  a newbie's point of view**. More experienced users will know how to tweak the tool to their own needs.

**Comment by sharkdp** [comment]
> Overall the question comes down to which is the worst case:
> 
>     1. A user can't find a file they expected to because it is captured by `.gitignore`
> 
>     2. A user gets additional output
> 
> 
> I think 1 is a obviously a _**significantly**_ worse situation. There's an adage in experimental sciences that applies: it is better to record too much information than too little. This is because too little has (generally) far higher costs than not enough (which requires doing everything over because you didn't know everything you'd need beforehand).

I think this is a really good point and I am seriously considering a switch of the default behavior in fd version 9. This would be a major breaking change. I know for a fact that people *are* using `fd` in scripts and pipelines. They will have to adapt (check) their code when upgrading to fd 9.

One practical problem is that we have a set of (short) command-line options that are designed to work with the current default, like `-I`/`--no-ignore`. We have a (somewhat hidden) `--ignore` counterpart, but no suitable short option. We would also have to figure out what to do with `--no-ignore-vcs`, `--no-ignore-parent`, `unrestricted`, etc.

**Comment by tmccombs** [comment]
I'm not entirely opposed to a change in the default, as long as there is an easy way for users to keep the current behavior if they want. Which could be as simple as being able to do `alias fd="fd --ignore-vcs"` (or `--ignore-git`), as long as I can still use `--no-ignore-vcs`, `-u`, `-I`, etc. to turn off the previous `--ignore-vcs` (which is how it currently works). 

> One practical problem is that we have a set of (short) command-line options that are designed to work with the current default, like -I/--no-ignore

That brings up the question, should the new default be the equivalent of current `fd --no-ignore`, or `fd --no-ignore-vcs`? 

Personally, I think it would be a little surprising if fd doesn't respect .fdignore files by default. .ignore  is more questionable. OTOH, in the case that you don't use any ignore files, bypassing the ignore machinery could improve performance.

If we changed the default to `fd --no-ignore-vcs`, then the `-I`, `--no-ignore` option would still be meaningful, since it excludes the .fdignore and .ignore files. Although perhaps not needed quite as often.

> We have a (somewhat hidden) --ignore counterpart, but no suitable short option.

For the long option, I think we would probably reverse the importance in the documentation (although maybe make the --no-ignore more prominent than it currently is). 

As for the short option, that depends on what direction we went with for --no-ignore vs --no-ignore-vcs as the default. 

If we went with `--no-ignore` as the default, `-i` would be a good choice as an alias for `--ignore`, except that it is already taken for "case-insensitive", but maybe we could change that, although that increases the potential breakage. Or we could invert the meaning of `-I`, which also would increase the scope of the breakage, and would be inconvenient for anyone who aliases fd to `fd --ignore`, since there isn't a short option to re-disable it, but maybe we could add a new short option for that as well. Or we could do something like `-I` means `--ignore` and `+I` means `--no-ignore`, but I don't think clap supports that convention, and it isn't a terribly common convention for CLIs.

If we went with `--no-ignore-vcs` as the default, there isn't currently a short option for `--no-ignore-vcs`, but it might be worth adding a short `--ignore-vcs`, perhaps `-G` for git? Although if we ever added support for additional VCSs that would make less sense.  `-v` is currently available, but I worry about that being confused for "version" or "verbose" (and possibly we would want to use that for an alias to `--verbose` at some point?).

> We would also have to figure out what to do with --no-ignore-vcs, --no-ignore-parent, unrestricted, etc.

I think those could probably stay the same as they are. Although make `--ignore-vcs` the main option documented instead of `--no-ignore-vcs`. 


> Maybe a minor update can be pushed to give a deprecation warning before any changes are made?

Where/when would we show this deprecation warning? Every time fd ran without a `--no-ignore(-vcs)` flag?  That would be incredibly annoying IMO.


> For the --no-ignore* flags, I think they should stay for a bit but just be non-operations (later you can leave but remove from man because it is just legacy support and add a deprecation flag here saying no longer needed)

No, we should keep them. Because I think we should support the use case of using an alias (or wrapper script) that passes `--ignore-*`, but allow negating it by `--no-ignore-*` later in the arg list. Just as we currently allow passing `--ignore` to undo a previous `--no-ignore`. 

> I don't think short flags are always necessary (I actually encourage people to use the log flags for aliases or scripts because they are better documentation), but I definitely get the desire

For scripts or aliases, I absolutely agree. However, for interactive use, I think that having short names for commonly used options is very valuable. And I think that turning the ignore functionality back on would be a pretty common usage, at least for me.

> rg is not part of the standard command set and isn't really relevant to this conversation.

fd uses the same code for determining which files to ignore as rg. Some of `fd`'s options were designed specifically to match options in `rg`. I generally view `fd` being to find what `rg` is to `grep`. And I strongly suspect that there is a large overlap between users of `rg` and users of `fd`.  I _do_ think it is relevant to the conversation. Maybe for searching for files  based on their names, respecting .gitignore is less important than it is for ripgrep. But if so, I think it is worth asking why that is.

**Comment by stevenwalton** [comment]
I think ignoring a `.fdignore` is perfectly acceptable as the default, and probably even good. Since these are files specifically for `fd`, just like it is perfectly reasonable for git to ignore `.gitignore` files by default. It's kinda self documentation. `fd` should also ignore hidden files by default because they are, after all... hidden and that's what is expected from a user. I'm not sure anything else should be ignored by default unless there's a convincing argument that someone who hasn't read the documentation would reasonably expect to ignore these files by default. I think you could argue something like a backup file `~*` or `.sw?` but I'd default to filtering only the minimal as the default option. Remember, users do have `grep` (and `sed`, `awk`, etc) and they're probably already used to filtering because `find` will spew out plenty of files at you and that's not a bug. 

So I think as long as the default options fit "what would a user that only has basic `find` knowledge expect" (because let's be real, most users don't know `find` very well either but there are the basic common patterns that are much more well known). Additional flags that apply different filters? I'm all for that but I'm also not the one putting in the time and effort so that's easy for me to say lol

**Comment by bingxiao** [comment]
Took 30min to finally figure out why fd cannot find a config file under /opt/homebrew which turned out being ignored by .gitignore, and why fd chose to design as such.

**Comment by stevenwalton** [comment]
The design choice is explained above. @sharkdp has explained, so please read.

I'm on your side of the decision btw, I think it is the better option and creates the least surprise (see discussion above). I think @sharkdp is open to the idea (it is marked as a milestone) **_but_** let's also be real that changing creates new surprises for others. `fd` is a great tool, even with this issue, and let's not pile on devs. They are sharing the tool with us, **for free**, and we're at a time where open source devs are under a lot more pressure. Let's not add on more, and ensure the conversation is productive. 

Btw, @bingxiao, heads up that `rg` has the same pattern. That one bit me **_really_** hard, but I think you'll have less success convincing the dev over there.

**Comment by bingxiao** [comment]
Please consider my previous post as a simple A/B vote. @stevenwalton I apologise the feedback came across as stressful. That was not my intention.

Appreciate 'find', 'git' and 'fd' for being helpful over the years. I temporarily compiled a local patched version of `fd` too. 

For me `rg -h` states explicitly about respecting gitignore by default so it mitigate the issue.

**Comment by stevenwalton** [comment]
@bingxiao , that's a fair point about the help file. I went ahead and created #2011 since it seems like a quick and easy "fix". That led to some of my initial confusion as well. Does that help string read better to you? @sharkdp would you take a look? 

(Gotta love GitHub... put in `#612` and tab corrects it to `#621`...)

**Comment by bingxiao** [comment]
@stevenwalton Thanks for your quick follow-up. That makes my mental model consistent when using the command. Appreciate it.
